### PR TITLE
chore: Move all server action to new authorization approach

### PR DESCRIPTION
**Pull request by gupta-piyush19**
WIP:
- To resolve merge conflicts ✅ 
- To test(e2e and functionality manually ✅ 

## What does this PR do?
Move all server action to new authorization approach

Fixes https://github.com/formbricks/internal/issues/294

## How should this be tested?

- Test A
- Test B

### Required

- [ ] Filled out the "How to test" section in this PR
- [ ] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [ ] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand bits
- [ ] Ran `pnpm build`
- [ ] Checked for warnings, there are none
- [ ] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [ ] My changes don't cause any responsiveness issues
- [ ] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/actions.ts [hunk 1]
```diff
@@ -2,106 +2,128 @@
...
+  surveyId: z.string(),
+});
+
+export const generateResultShareUrlAction = authenticatedActionClient
+  .schema(ZGenerateResultShareUrlAction)
+  .action(async ({ ctx, parsedInput }) => {
+    await checkAuthorization({
+      userId: ctx.user.id,
+      organizationId: await getOrganizationIdFromSurveyId(parsedInput.surveyId),
+      rules: ["survey", "update"],
```
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/actions.ts:55]: Let's make two auth checks here: `"response", "update"` + `"survey", "update"`
It's not perfect, but given users who have the right to update the survey and the responses also the authority to distribute the results and update the survey object, seems like the closest fit.

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/actions.ts [hunk 2]
```diff
@@ -2,106 +2,128 @@
...
+  surveyId: z.string(),
+});
+
+export const getResultShareUrlAction = authenticatedActionClient
+  .schema(ZGetResultShareUrlAction)
+  .action(async ({ ctx, parsedInput }) => {
+    await checkAuthorization({
+      userId: ctx.user.id,
+      organizationId: await getOrganizationIdFromSurveyId(parsedInput.surveyId),
+      rules: ["survey", "read"],
```
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/actions.ts:83]: `"response", "read"`

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/actions.ts [hunk 3]
```diff
@@ -2,106 +2,128 @@
...
+  surveyId: z.string(),
+});
+
+export const deleteResultShareUrlAction = authenticatedActionClient
+  .schema(ZDeleteResultShareUrlAction)
+  .action(async ({ ctx, parsedInput }) => {
+    await checkAuthorization({
+      userId: ctx.user.id,
+      organizationId: await getOrganizationIdFromSurveyId(parsedInput.surveyId),
+      rules: ["survey", "update"],
```
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/actions.ts:104]: `"response", "update"` + `"survey", "update"`

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/actions.ts [hunk 4]
```diff
@@ -1,53 +1,67 @@
...
 
-  const isAuthorized = await canUserAccessSurvey(session.user.id, surveyId);
-  if (!isAuthorized) throw new AuthorizationError("Not authorized");
+export const getResponsesDownloadUrlAction = authenticatedActionClient
+  .schema(ZGetResponsesDownloadUrlAction)
+  .action(async ({ ctx, parsedInput }) => {
+    await checkAuthorization({
+      userId: ctx.user.id,
+      organizationId: await getOrganizationIdFromSurveyId(parsedInput.surveyId),
+      rules: ["survey", "read"],
```
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/actions.ts:25]: `"response", "read"`

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/actions.ts [hunk 5]
```diff
@@ -1,53 +1,67 @@
...
-};
+    return getResponseDownloadUrl(parsedInput.surveyId, parsedInput.format, parsedInput.filterCriteria);
+  });
 
-export const getSurveyFilterDataAction = async (surveyId: string, environmentId: string) => {
-  const session = await getServerSession(authOptions);
-  if (!session) throw new AuthorizationError("Not authorized");
+const ZGetSurveyFilterDataAction = z.object({
+  surveyId: z.string(),
+  environmentId: z.string(),
```
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/actions.ts:33]: Can we derive the `environmentId` from the `survey` to make sure we don't have inconsistencies?

File: packages/ee/multi-language/lib/actions.ts [hunk 6]
```diff
@@ -1,78 +1,90 @@
...
+      userId: ctx.user.id,
+      organizationId: await getOrganizationIdFromEnvironmentId(parsedInput.environmentId),
+      rules: ["language", "delete"],
+    });
+
+    return await deleteLanguage(parsedInput.environmentId, parsedInput.languageId);
+  });
+
+const ZGetSurveysUsingGivenLanguageAction = z.object({
+  productId: z.string(),
```
Comment (mattinannt) [inline packages/ee/multi-language/lib/actions.ts:56]: SECURITY RISK: We should derive the productId from the languageId, otherwise a malicious user can get the surveys for a languageId from an org where he doesn't have access by providing a productId from his org.

File: packages/lib/response/service.ts [hunk 7]
```diff
@@ -95,6 +95,12 @@ export const responseSelection = {
       },
       isResolved: true,
       isEdited: true,
+      response: {
```
Comment (mattinannt) [inline packages/lib/response/service.ts:98]: why is that needed? we already fetch the response id and surveyId

File: packages/types/action-client.ts [hunk 8]
```diff
@@ -9,6 +9,18 @@ export const ZResource = z.enum([
...
+  "responseNote",
+  "membership",
+  "attributeClass",
+  "segment",
+  "actionClass",
+  "integration",
+  "webhook",
+  "apiKey",
+  "subscription",
+  "invite",
```
Comment (mattinannt) [inline packages/types/action-client.ts:22]: I'm not sure if `invite` and `membership` needs to be different. When people have `membership, create` rights that also means that they need to be able to create an invite.

**Review by mattinannt — CHANGES_REQUESTED** [review]
@gupta-piyush19 Thanks a lot for the (very long) PR; great job!!! 🚀🔥
Due to the amount of actions and services touched in this PR I found a few inconsistencies in the services which make sure to solve here I think since they got highlighted through the updated actions.

Also I think for the action types every time an `id` (e.g. `environmentId`) is used, we can narrow down the types more to use the type `ZId` to make sure it's a CUID, instead of just using `z.string()`

File: packages/lib/actionClient/permissions.ts [hunk 9]
```diff
@@ -2,69 +2,270 @@ import { ZProductUpdateInput } from "@formbricks/types/product";
 
 export const Permissions = {
   owner: {
+    environment: {
+      read: true,
```
Comment (mattinannt) [inline packages/lib/actionClient/permissions.ts:6]: Not sure if this permission makes sense, since every user needs to be able to read the environment, otherwise using formbricks is impossible 😉

File: packages/lib/actionClient/permissions.ts [hunk 10]
```diff
@@ -74,37 +275,166 @@ export const Permissions = {
...
+      create: false,
+      delete: false,
+    },
+    integration: {
+      create: false,
+      delete: false,
+    },
+    webhook: {
+      create: false,
+      update: true,
```
Comment (mattinannt) [inline packages/lib/actionClient/permissions.ts:414]: This should be false

File: apps/web/app/(app)/environments/[environmentId]/settings/(organization)/members/actions.ts [hunk 11]
```diff
@@ -46,177 +37,189 @@ export const updateOrganizationNameAction = authenticatedActionClient
...
-  if (!session) {
-    throw new AuthenticationError("Not authenticated");
-  }
+export const createInviteTokenAction = authenticatedActionClient
+  .schema(ZCreateInviteTokenAction)
+  .action(async ({ parsedInput, ctx }) => {
+    await checkAuthorization({
+      userId: ctx.user.id,
+      organizationId: await getOrganizationIdFromInviteId(parsedInput.inviteId),
+      rules: ["invite", "read"],
```
[the remark this answers is not in the document]
Comment (gupta-piyush19) [reply apps/web/app/(app)/environments/[environmentId]/settings/(organization)/members/actions.ts:117]: I think `rules: ["invite", "create"]` makes more sense, as we're creating an invite token for the invite, though I chose permissions based on the operations we were doing on the entity. As in this case, we sign a new JWT token, if the invite is valid(after reading the invite)

File: packages/lib/actionClient/permissions.ts (hunk 9 above)
Comment (gupta-piyush19) [reply packages/lib/actionClient/permissions.ts:6]: I added this to make the `checkAuthorization` check consistent. This is more of a redundant check, but organizationId is also being checked based on the environment when this permission is being used.

File: packages/types/action-client.ts (hunk 8 above)
Comment (gupta-piyush19) [reply packages/types/action-client.ts:22]: Yeah, the meaning is identical mostly, but since they possess different schemas and may have more granular control over them (for updates, etc.), I added permissions for other entities.

File: apps/web/app/(app)/environments/[environmentId]/settings/(organization)/members/actions.ts [hunk 12]
```diff
@@ -46,177 +37,189 @@ export const updateOrganizationNameAction = authenticatedActionClient
...
-  if (!session) {
-    throw new AuthenticationError("Not authenticated");
-  }
+export const leaveOrganizationAction = authenticatedActionClient
+  .schema(ZLeaveOrganizationAction)
+  .action(async ({ parsedInput, ctx }) => {
+    await checkAuthorization({
+      userId: ctx.user.id,
+      organizationId: parsedInput.organizationId,
+      rules: ["membership", "delete"],
```
[the remark this answers is not in the document]
Comment (gupta-piyush19) [reply apps/web/app/(app)/environments/[environmentId]/settings/(organization)/members/actions.ts:86]: Ahh, nice! good catch 😲 
Changed the permission to `rules: ["organization", "read"]`

File: packages/lib/response/service.ts (hunk 7 above)
Comment (gupta-piyush19) [reply packages/lib/response/service.ts:98]: I added this earlier as I got some errors in the types inference. To fix this, I've extended the select value for the `getResponseNote` function, ensuring that other data-fetching calls remain the same.

**Comment by gupta-piyush19** [comment]
Thanks, @mattinannt, for the review; I've fixed most of them already and commented on a few. Can you please re-review it?

**Review by mattinannt — APPROVED** [review]
@gupta-piyush19 thanks a lot for making the changes 😊 Looks great and is ready to merge! 🚀🚀🚀🔥
